ome/ome-model
# Add gradle build

--exclude

This PR adds a gradle build system to replace the existing maven build.  Compatible with #65. The maven build is left in place for now to avoid conflicts. CI tests are now using gradle.  gradle and openjdk7 aren't compatible on travis (see https://github.com/travis-ci/travis-ci/issues/8503), so test with 8 and 10, dropping 7.  The source version could be upgraded to 1.8, but this has been left for a future PR.

This should be a drop-in replacement for all of the maven functionality in current use, and it's vastly shorter and simpler than the maven logic.  This is the most complex bioformats-related package, with multiple components, source code generation and documentation generation, so this conversion is as complex as it will get for all of the decoupled components.  It includes all the needed logic for building, testing and making signed releases.

Testing:

Run `./gradlew build|assemble|test|publishToMavenLocal`

To test packaging changes:

```
mvn
./gradlew assemble
pkgdiff ome-xml/target/ome-xml-5.6.4-SNAPSHOT.jar ome-xml/build/libs/ome-xml-5.6.4-SNAPSHOT.jar
(repeat for other jars, and look at the HTML pages for changes; should be cosmetic only--pom.xml/pom.properties not in jar [maven-specific] and some minor MANIFEST.MF differences)
```

Workflow changes:

- Edit `VERSION` to set the version number.  Used by both the gradle and cmake builds.
- Create `~/.gradle/gradle.properties` (equivalent of `~/.m2/settings.xml`) with this content:

```
signing.keyId=<keyid>
ossrhUsername=<username>
ossrhPassword=<password>
```

- Run `./gradlew publish` to upload to maven central.  Snapshots to snapshots, releases to staging handled automatically.  Or run `./gradlew publishSnapshot` to upload snapshots only, skipping releases, which is useful for CI builds where signing is not possible.
